src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php

```diff
@@ -0,0 +1,210 @@
+        unset($amqpOptions['exchange']);
+
+        return new self($amqpOptions, $exchangeOptions, $queueOptions, $debug, $amqpFactory);
+    }
+
+    /**
+     * @throws \AMQPException
+     */
+    public function publish(string $body, array $headers = array())
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $this->exchange()->publish($body, null, AMQP_NOPARAM, array('headers' => $headers));
+    }
+
+    /**
+     * Waits and gets a message from the configured queue.
+     *
+     * @throws \AMQPException
+     */
+    public function waitAndGet(): ?\AMQPEnvelope
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $message = null;
+
+        try {
+            $this->queue()->consume(function (\AMQPEnvelope $envelope) use (&$message): bool {
+                $message = $envelope;
+
+                return false;
+            });
+        } catch (\AMQPQueueException $e) {
+            if (404 === $e->getCode()) {
+                // If we get a 404 for the queue, it means we need to setup the exchange & queue.
+                $this->setup();
+
+                return $this->waitAndGet();
+            }
+
+            throw $e;
+        }
+
+        return $message;
+    }
+
+    private function channel(): \AMQPChannel
+    {
+        if (null === $this->amqpChannel) {
+            $connection = $this->amqpFactory->createConnection($this->connectionCredentials);
+            $connectMethod = 'true' === ($this->connectionCredentials['persistent'] ?? 'false') ? 'pconnect' : 'connect';
+
+            if (false === $connection->{$connectMethod}()) {
+                throw new \AMQPException('Could not connect to the AMQP server. Please verify the provided DSN.');
+            }
+
+            $this->amqpChannel = $this->amqpFactory->createChannel($connection);
+        }
+
+        return $this->amqpChannel;
+    }
+
+    private function queue(): \AMQPQueue
+    {
+        if (null === $this->amqpQueue) {
+            $this->amqpQueue = $this->amqpFactory->createQueue($this->channel());
+            $this->amqpQueue->setName($this->queueConfiguration['name']);
+            $this->amqpQueue->setFlags($this->queueConfiguration['flags'] ?? AMQP_DURABLE);
+        }
+
+        return $this->amqpQueue;
+    }
+
+    private function exchange(): \AMQPExchange
+    {
+        if (null === $this->amqpExchange) {
+            $this->amqpExchange = $this->amqpFactory->createExchange($this->channel());
+            $this->amqpExchange->setName($this->exchangeConfiguration['name']);
+            $this->amqpExchange->setType($this->exchangeConfiguration['type'] ?? AMQP_EX_TYPE_FANOUT);
+            $this->amqpExchange->setFlags($this->exchangeConfiguration['flags'] ?? AMQP_DURABLE);
+        }
+
+        return $this->amqpExchange;
+    }
+
+    public function ack(\AMQPEnvelope $message)
+    {
+        return $this->queue()->ack($message->getDeliveryTag());
+    }
```

Comment:
we group methods by visibility, public before private (same for properties/consts)

Reply:
Changed :+1: